File: Assets/Scripts/Utils/StateMachine.cs
using System;
using System.Collections.Generic;
using UnityEngine;

namespace MidnightReturn.Utils
{
    public abstract class State<T>
    {
        protected T Owner;
        public string Name { get; protected set; }

        protected State(T owner, string name) { Owner = owner; Name = name; }

        public virtual void OnEnter(State<T> previous) { }
        public virtual void OnUpdate(float dt) { }
        public virtual void OnFixedUpdate(float dt) { }
        public virtual void OnExit(State<T> next) { }
    }

    public class StateMachine<T>
    {
        private readonly Dictionary<string, State<T>> _states = new();
        private State<T> _current;
        private bool _locked;

        public string CurrentState => _current?.Name ?? "none";
        public bool Is(string name) => _current?.Name == name;

        public StateMachine<T> Add(State<T> state)
        {
            _states[state.Name] = state;
            return this;
        }

        public bool Transition(string name)
        {
            if (_locked || !_states.TryGetValue(name, out var next)) return false;
            if (next == _current) return false;

            var prev = _current;
            prev?.OnExit(next);
            _current = next;
            _current.OnEnter(prev);
            return true;
        }

        public void Update(float dt) => _current?.OnUpdate(dt);
        public void FixedUpdate(float dt) => _current?.OnFixedUpdate(dt);
        public void Lock()   => _locked = true;
        public void Unlock() => _locked = false;
    }
}

File: web-prototype/src/utils/StateMachine.ts
export interface IState {
  name: string;
  onEnter?(prevState?: IState): void;
  onUpdate?(dt: number): void;
  onExit?(nextState?: IState): void;
}

export class StateMachine {
  private states = new Map<string, IState>();
  private current: IState | null = null;
  private locked = false;

  add(state: IState): this {
    this.states.set(state.name, state);
    return this;
  }

  transition(name: string): boolean {
    if (this.locked) return false;
    const next = this.states.get(name);
    if (!next || next === this.current) return false;

    const prev = this.current;
    this.current?.onExit?.(next);
    this.current = next;
    this.current.onEnter?.(prev ?? undefined);
    return true;
  }

  update(dt: number): void {
    this.current?.onUpdate?.(dt);
  }

  get state(): string {
    return this.current?.name ?? 'none';
  }

  is(name: string): boolean {
    return this.current?.name === name;
  }

  lock(): void  { this.locked = true; }
  unlock(): void { this.locked = false; }
}
